src/search/algorithms/named_vector.h

```diff
@@ -0,0 +1,103 @@
+#ifndef ALGORITHMS_NAMED_VECTOR_H
+#define ALGORITHMS_NAMED_VECTOR_H
+
+#include <string>
+#include <vector>
+
+namespace named_vector {
+/*
+  NamedVector is a vector-like collection with optional names to be associated with each element.
+  It is optimized to have minimal overhead when there are no names.
+  Any name which is not specified is assumed to be an empty string.
+ */
+template<typename T>
+class NamedVector {
+    std::vector<T> elements;
+    std::vector<std::string> names;
+public:
+    NamedVector() = default;
```

Comment:
Since we do not change any of the constructors, we should leave out all of them.

This is called the "rule of zero" and is a bit of a controversial topic apparently (https://www.fluentcpp.com/2019/04/23/the-rule-of-zero-zero-constructor-zero-calorie/).

Reply:
I don't think the rule of zero is that controversial, or else it wouldn't be so prominent in the Core Guidelines. Of course there are different opinions by someone on *anything*, but I find the discussed Scott Meyer example unconvincing, at least from the second-hand discussion. Firstly, it shows that there is an issue if you *violate* the rule of zero by introducing an explicit destructor, but of course the point of the rule is to follow it, not violate it. And these issues are caught by compiler warnings.

But this is beside the main point. The recommendations are either the rule of zero or the rule of five. The code we currently have is exactly the kind of code Scott Meyer warns against when misusing the rule of 0 because (if I see it correctly) it implicitly defines move assignment as copy assignment. If I understand Meyer's argument correctly, he thinks the rule of 5 should be followed rather than the rule of 0 because if you start with the rule of 0 and then (by not paying attention) add *some* of the 5 methods but not all, you add in a dangerous spot. This is exactly where the current version of the code is.

Following either the rule of 0 or the rule of 5 would fix that. I don't feel very strongly about which of the two we follow, but the rule of 0 certainly seems to be what the C++ community is moving towards and what we have started using more frequently in the code.

Reply:
I think "controversial" may have come across in the wrong way: I just meant the different recommendations of Meyers and the core guidelines. I didn't want to imply that we are not agreeing on what rule to follow.

Reply:
PS: Also the page that Florian linked ended by saying that this is not controversial:

> To conclude I’ll leave the closing word to Arne Mertz, that summarized the debate with a rule that everyone agrees on, the “Rule of All or Nothing”:
>    As long as you can, stick to the Rule of Zero, but if you have to write at least one of the Big Five, default the rest.

...and this is kind-of what the Core Guidelines say as well (C.20 + C.21), although C.21 doesn't explicitly mention "defaulting" but rather "defining". But defaulting is of course the simplest way to define something where it has the right semantics.